### PR TITLE
fix(augmentation): return a copy from CenterCrop slice mode

### LINKED ISSUE #4413: W1b: `CenterCrop(cropping_mode="slice")` returns a view that aliases its input
`CenterCrop` in its default `cropping_mode="slice"` returns a view of the caller's tensor rather than a copy. The output shares storage with the input, so writing into the augmentation's result silently corrupts the tensor the caller passed in — at the offset the crop was taken from, which is not a position the caller can predict without reconstructing the crop box.

The other two crop classes in the same mode do not do this. `RandomCrop` and `RandomResizedCrop` return copies, including a `RandomResizedCrop` whose crop is not resized afterwards, and `CenterCrop` in `cropping_mode="resample"` also returns a copy. So the three classes disagree about a property a caller has no reason to check per class, and the divergence is not documented on any of them.

This matters beyond in-place writes: a view keeps the whole original tensor alive, so a pipeline that crops a large batch down and keeps the crops holds the full input in memory indefinitely.

## Repro

```python
import torch, kornia, kornia.augmentation as K
print("kornia", kornia.__version__, "torch", torch.__version__)
for cls, kwargs in ((K.CenterCrop, {"size": 4}),
                    (K.RandomCrop, {"size": (4, 4)}),
                    (K.RandomResizedCrop, {"size": (4, 4), "scale": (0.25, 0.25)})):
    torch.manual_seed(0)
    x = torch.arange(48., dtype=torch.float32).reshape(1, 1, 6, 8)
    aug = cls(cropping_mode="slice", p=1.0, **kwargs)
    o = aug(x)
    shares = o.untyped_storage().data_ptr() == x.untyped_storage().data_ptr()
    o[0, 0, 0, 0] = -99.0
    print(f"{cls.__name__:18s} slice: is_view={o._base is not None}  shares_storage={shares}"
          f"  input_corrupted={bool((x == -99.0).any())}  at={[t.tolist() for t in (x == -99.0).nonzero()]}")
torch.manual_seed(0)
x = torch.arange(48., dtype=torch.float32).reshape(1, 1, 6, 8)
o = K.CenterCrop(4, cropping_mode="resample", p=1.0)(x); o[0, 0, 0, 0] = -99.0
print("CenterCrop         resample: input_corrupted=", bool((x == -99.0).any()))
```

## Observed

```
kornia 0.9.0rc1 torch 2.14.0
CenterCrop         slice: is_view=True  shares_storage=True  input_corrupted=True  at=[[0, 0, 1, 2]]
RandomCrop         slice: is_view=False  shares_storage=False  input_corrupted=False  at=[]
RandomResizedCrop  slice: is_view=False  shares_storage=False  input_corrupted=False  at=[]
CenterCrop         resample: input_corrupted= False
```

The write to the output's `(0, 0)` landed at the input's `(1, 2)` — the top-left corner of the crop box on this 6 × 8 input.

## Expected

An augmentation returns a tensor the caller can write to without corrupting the input. `CenterCrop`'s `slice` branch should return a contiguous copy, matching `RandomCrop` and `RandomResizedCrop` in the same mode and `CenterCrop` itself in `resample` mode.

That is a focused fix and welcome as a PR. It has a measurable cost — `slice` mode exists to avoid a resample, and a copy is not free — so if the aliasing is to be kept as a deliberate optimisation, the alternative is to document it on `CenterCrop` and to make the other two classes behave the same way, which is the larger change of the two.

One caveat for whoever takes it: at `p=0.0` the module hands the input straight back, so identity of storage is not by itself evidence of the crop branch being taken.

## Classification

**Broken output** — writing to the augmentation's output mutates the caller's input tensor, silently and at an unpredictable offset. Idiom: 6 (silently mutating an input — here by handing out a writable alias of it).

## Disposition

`MERGE IN WINDOW — behavior change`. Returning a copy changes the aliasing contract that any current caller relying on the view depends on, and it changes the memory profile of `slice` mode. Treating the aliasing as a deliberate variant is not available either, because the divergence from the other two crop classes is unintentional. Batch 6 documents the aliasing on `CenterCrop` and pins it; it does not change it.

## Environment

kornia 0.9.0rc1 @ `d6a2c138` (main), torch 2.14.0, python 3.11, cpu. Found during the augmentation conventions audit (batch 6, tracking issue #4407); the batch documents the behavior above as it is and pins it, and does not change it.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by SulimanAbdulrazzaq**
## What does this pull request do?

In its default `cropping_mode="slice"`, `CenterCrop` returned `input[..., top:top + h, left:left + w]`. That is a view of the caller's tensor, so writing to the augmentation's output changed the input at the crop offset. `RandomCrop` and `RandomResizedCrop` already return new tensors in slice mode, because in eager mode `crop_by_indices` fills a new tensor. So does `CenterCrop` in `resample` mode.

The slice branch now returns a clone of the slice. `.contiguous()` would not be enough. A full-size crop is already contiguous, and so is a crop that keeps whole rows of a `(1, 1, H, W)` input, so either would come back as the same view.

The Convention block no longer describes the view as a wart tracked in #4413. `test_wart_center_crop_slice_aliases_input_4413` becomes `test_convention_center_crop_slice_returns_a_copy_4413`. The new test covers three crops of a `6 x 8` input: a strided `(4, 4)` crop, a rows-only `(4, 8)` crop and a full-size `(6, 8)` crop. For each, it checks that the output has its own storage and that filling the output leaves the input unchanged.

This follows the first option in the issue and returns a copy. The alternative, documenting the view and making the other two classes alias too, is not part of this change. The copy is one allocation the size of the crop, which `RandomCrop` and `RandomResizedCrop` already make in slice mode.

The changelog fragment `changelog.d/<PR>.fixed.md` follows in a separate commit once this PR has a number.

## Related issue or discussion

Fixes #4413

## What did you check?

Windows 11, Python 3.11.0, torch 2.14.0+cpu, kornia installed in editable mode.

```text
$ python -m pytest tests/augmentation/test_conventions_geometric_crop.py -k 4413 --device=cpu --dtype=float32,float64,float16,bfloat16
new test with the main source: 12 failed      this PR: 12 passed

$ python -m pytest tests/augmentation -k "CenterCrop or center_crop or crop" --device=cpu --dtype=float32,float64 -n 2
main:    554 passed, 48 skipped, 5 xfailed, 2 xpassed
this PR: 558 passed, 48 skipped, 5 xfailed, 2 xpassed   (the old wart test ran twice, the new test runs six times)

$ python -m pytest tests/augmentation --device=cpu --dtype=float32 -n 2
3902 passed, 3403 skipped, 31 xfailed, 5 xpassed

$ python -m pytest tests/augmentation/test_torch_export.py -k CenterCrop --device=cpu --dtype=float32
3 passed

$ python -m pytest --doctest-modules kornia/augmentation/_2d/geometric/center_crop.py
1 passed

$ pre-commit run --files kornia/augmentation/_2d/geometric/center_crop.py tests/augmentation/test_conventions_geometric_crop.py
all hooks Passed (ruff check, ruff format, codespell, license header, whitespace/EOF/line endings)

$ ty check kornia    # ty 0.0.9
All checks passed!
```

`TestCenterCrop::test_dynamo` could not run here. Inductor raises `InvalidCxxCompiler: Compiler: cl is not found` on this Windows machine, on `main` and on this branch alike. As a substitute, `K.CenterCrop(size)` was compiled with `torch.compile(..., fullgraph=True, backend="aot_eager")` on a `(2, 3, 20, 24)` input for sizes `(16, 10)`, `(20, 24)` and `(4, 24)`. It compiles without a graph break and matches eager output on both `main` and this branch. The compiled output shares the input's storage on `main` and has its own storage here.

## How did AI help?

This change was made with Claude Code, Anthropic's AI coding agent. It traced the aliasing to the slice branch, compared it with the other crop paths, wrote the fix and the test, and ran the checks listed above. Before this PR was opened, a second, independent Claude Code session reviewed the diff and re-ran the new test on `main` and on this branch, the crop subset and the doctest in a clean checkout. The results in this description come from those runs.

### CODE REVIEW

**Review by ducha-aiki — CHANGES_REQUESTED** [review]
Review of `9dfe0e0ee`

The PR makes the `cropping_mode="slice"` branch of `CenterCrop.apply_transform` return `input[..., crop].clone()` instead of the view. It rewrites the Convention block, which described the view as the #4413 wart, and inverts `test_wart_center_crop_slice_aliases_input_4413` into `test_convention_center_crop_slice_returns_a_copy_4413`. The new test is parametrized over a strided `(4, 4)` crop, a rows-only `(4, 8)` crop and a full-size `(6, 8)` crop. The code change is correct and complete. I have one changelog ask.

**What I verified** (torch 2.14 CPU unless noted; `kornia.__file__` printed for each tree)

- Storage sharing between output and input, origin/main vs this PR. The PR only changes the crop branch: `p=0` and a `p_batch` draw that skips the crop still hand back the input tensor itself, as every augmentation does (the caveat noted in #4413).

  | case | main | PR |
  |---|---|---|
  | `(B,C,H,W)`, sizes `(4,4)`, `(4,8)`, `(2,8)`, `(6,8)`, `keepdim` False/True | shared | own storage |
  | `(C,H,W)` input, `keepdim=True`, full-size crop | shared | own storage |
  | `p=0.5`, crop applied | shared | own storage |
  | `AugmentationSequential(..., data_keys=["input", "mask"])`, input and mask | shared | own storage |
  | `channels_last` input | shared | own storage, still `channels_last` |
  | `torch.compile(fullgraph=True)`, `(16, 10)` and `(20, 24)` | shared | own storage, `torch.equal` to eager |

- Other properties. Crop values are unchanged, because the slice is the same one. The gradient is still the indicator of the crop box. `gradcheck` (float64) passes. The dtype is preserved for float16, bfloat16 and float64. `resample` mode and `inverse` are untouched. `CenterCrop3D` and `RandomCrop3D` already return copies, so nothing is left to fix in 3D.
- Mutations of the library line. Dropping `.clone()` gives 3/3 red. `.clone()` → `.contiguous()` gives 2/3 red: the rows-only and full-size rows catch it, as the PR description says they should. `.clone()` → `.reshape(...)` gives 3/3 red. On origin/main library code the new test fails 6/6 (float32, float16).
- The only `#4413` hits in `kornia/`, `tests/` and `docs/` are the new code comment and the test comment. The old wart pin is inverted. There is no remaining "view", "alias" or "shares storage" prose in `center_crop.py`, `conventions.rst` or the crop conventions tests.
- Test runs. On the PR head, `test_conventions_*.py` plus the `CenterCrop` tests in `test_augmentation.py` give 1160 passed. The inductor `TestCenterCrop` dynamo test passes. On the merge tree with origin/main (`d0b8328e2`, merges cleanly), `test_conventions_*.py` plus `test_torch_export.py` (float32, float64) give 1133 passed, and `tests/augmentation -k CenterCrop` (float32, float16, bfloat16) gives 141 passed. On the torch 2.5.1 floor, the `CenterCrop` and #4413 tests give 21 passed.
- CI. Everything is green except `half-precision (float16) / tests (3.11, 2.9.1)` (cancelled) and `collector`. Those two are the known Intel-host hang, #4214, and are not caused by this PR.

**P2: the fragment should be `breaking`, not only `fixed`**

The code change is right. What it does to users, though, is an incompatible change to the contract that has shipped, and kornia has recorded exactly this kind of change as breaking before:

- #4500 made `PinholeCamera(...)` clone its inputs instead of storing them by reference. It shipped `4500.breaking.md`: "code that relied on bidirectional storage aliasing must mutate the camera's tensors explicitly".
- The disposition on #4413 reads `MERGE IN WINDOW — behavior change`: "Returning a copy changes the aliasing contract that any current caller relying on the view depends on, and it changes the memory profile of `slice` mode."

Two things change for callers:

- Code that writes into the input after the crop (`out = CenterCrop(s)(x); x.mul_(k)`) and expects `out` to follow now silently reads stale values.
- `slice` mode now allocates one crop-sized tensor per call.

The Convention block labelled the view a wart. That is why this is a fix and needs no deprecation. It does not make the change invisible to a caller who depended on the view. Please add `changelog.d/4727.breaking.md` that states both the old and the new behaviour. Keep `4727.fixed.md` or fold it into the breaking fragment. For example:

> `CenterCrop` in the default `cropping_mode="slice"` now returns a copy of the cropped region. Previously it returned a view that shared storage with the input: writing to the output changed the input at the crop offset, and writing to the input after the call changed the output. Code that relied on either direction of that aliasing must now write to the tensor it wants changed. The crop now allocates one crop-sized tensor per call, as `RandomCrop` and `RandomResizedCrop` already do in slice mode, and no longer keeps the whole input alive through the output. Fixes #4413.

Nothing else is open. Once the breaking fragment is in, this is ready to merge.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by SulimanAbdulrazzaq** [comment]
Thanks for the thorough review. Done in 1504d55: `4727.fixed.md` is folded into `changelog.d/4727.breaking.md`, using your wording, and `.github/scripts/check_changelog.py` passes for this PR.

**Review by ducha-aiki — APPROVED** [review]
## Re-review at `1504d5531ac46917ac3e84bb57afa9ab5a486c65`

The one open item from the previous round is done. Approving.

- **Changelog.** `changelog.d/4727.fixed.md` is now `changelog.d/4727.breaking.md`. The fragment says what used to happen: slice mode returned a view, and writes went both ways, from the output into the input and from the input into the output. It says what happens now: a copy, one crop-sized allocation per call, and the output no longer keeps the input alive. `check_changelog.py --base-ref origin/main --pr-number 4727` reports the fragments as valid.
- **Delta since `9dfe0e0ee`.** The only change is the fragment rename and its new text. The library and test diff against the merge base is the same one I reviewed last round.
- **Merged with main (`8b03a8085`).** The merge is clean. With `kornia.__file__` resolving to the merge worktree:
  - `tests/augmentation -k "crop or Crop" --dtype=float32,float64`: 574 passed, 47 skipped, 5 xfailed, 2 xpassed.
  - The `-k 4413` pins pass in fp32, fp64, fp16 and bf16 (12 passed).
- **Mutation.** Dropping the `.clone()` makes 3 tests fail in the crop subset.
- **Aliasing.** `CenterCrop`, `RandomCrop` and `RandomResizedCrop` all return tensors that do not share storage with the input, which confirms the fragment's comparison.
- **Closing grep.** `#4413|issues/4413` over `kornia/ tests/ docs/` finds only the new code comment, the convention pin and the fragment. No stale wart prose is left.
- **CI.** 41/41 checks pass.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
